File: ipcam_bridge.cpp
// ipcam_bridge.cpp - FIXED VERSION
#include <opencv2/opencv.hpp>
#include <iostream>
#include <chrono>
#include <thread>
#include <fstream>
#include <iomanip>
#include <sstream>

int main() {
    std::string ip_url = "http://10.35.20.50:8080/video";
    
    cv::VideoCapture cap(ip_url);
    if (!cap.isOpened()) {
        std::cout << "Cannot open IP camera: " << ip_url << std::endl;
        return -1;
    }
    
    std::cout << "IP Camera opened successfully!" << std::endl;
    
    // Create directory for frames
    system("mkdir -p ipcam_frames");
    
    // Create timestamp file (like TUM dataset format)
    std::ofstream timestamp_file("rgb.txt");
    timestamp_file << "# color images\n";
    timestamp_file << "# file: 'ipcam_frames'\n";
    timestamp_file << "# timestamp filename\n";
    
    int frame_count = 0;
    cv::Mat frame;
    auto start_time = std::chrono::steady_clock::now();
    
    while (true) {
        cap >> frame;
        if (frame.empty()) {
            std::cout << "Empty frame received, retrying..." << std::endl;
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
            continue;
        }
        
        // Calculate timestamp in seconds
        auto current_time = std::chrono::steady_clock::now();
        double timestamp = std::chrono::duration<double>(current_time - start_time).count();
        
        // Pad frame count with zeros (6 digits)
        std::stringstream ss;
        ss << std::setw(6) << std::setfill('0') << frame_count;
        std::string frame_num = ss.str();
        
        // Save frame
        std::string filename = "ipcam_frames/" + frame_num + ".png";
        cv::imwrite(filename, frame);
        
        // Write to timestamp file
        timestamp_file << std::fixed << std::setprecision(6) << timestamp << " " << filename << "\n";
        
        // Show preview
        cv::imshow("Phone Camera", frame);
        
        std::cout << "Frame " << frame_count << " saved at " << timestamp << "s" << std::endl;
        
        frame_count++;
        
        // Break on ESC key or after 1000 frames
        char key = cv::waitKey(30);
        if (key == 27 || frame_count >= 1000) {  // ESC key
            break;
        }
    }
    
    cap.release();
    cv::destroyAllWindows();
    timestamp_file.close();
    
    std::cout << "\nSaved " << frame_count << " frames to ipcam_frames/" << std::endl;
    std::cout << "You can now run ORB-SLAM3 on these frames!" << std::endl;
    
    return 0;
}

File: ipcam_bridge.py
# ipcam_bridge.py
import cv2
import time
import os

# Create directory
os.makedirs("ipcam_frames", exist_ok=True)

# Open IP camera
cap = cv2.VideoCapture("http://10.35.20.50:8080/video")

if not cap.isOpened():
    print("Cannot open IP camera")
    exit()

print("IP Camera opened successfully!")

# Create timestamp file
timestamp_file = open("rgb.txt", "w")
timestamp_file.write("# color images\n")
timestamp_file.write("# file: 'ipcam_frames'\n")
timestamp_file.write("# timestamp filename\n")

frame_count = 0
start_time = time.time()

while True:
    ret, frame = cap.read()
    if not ret:
        print("Empty frame received, retrying...")
        time.sleep(0.1)
        continue
    
    # Calculate timestamp
    timestamp = time.time() - start_time
    
    # Save frame
    filename = f"ipcam_frames/{frame_count:06d}.png"
    cv2.imwrite(filename, frame)
    
    # Write to timestamp file
    timestamp_file.write(f"{timestamp:.6f} {filename}\n")
    
    # Show preview
    cv2.imshow("Phone Camera", frame)
    
    print(f"Frame {frame_count} saved at {timestamp:.2f}s")
    
    frame_count += 1
    
    # Break on ESC key or after 500 frames
    key = cv2.waitKey(30) & 0xFF
    if key == 27 or frame_count >= 500:  # ESC key
        break

cap.release()
cv2.destroyAllWindows()
timestamp_file.close()

print(f"\nSaved {frame_count} frames to ipcam_frames/")
print("You can now run ORB-SLAM3 on these frames!")